### ISSUE TITLE
Incorrect syntax highlight on coffee-script regexp

### ISSUE THREAD

**Issue by Anteru** [pygments/pygments-migration-test#698]
(Original issue 991 created by mirekrusin on 2014-05-13T19:40:31.934129+00:00)

```
#!coffeescript

foo = (str) ->
  /'|"/.test str

```

seems to break as seen on http://pygments.org/demo/378944/ or https://github.com/atom/bracket-matcher/blob/master/lib/bracket-matcher.coffee#L127-L148

**Comment by Anteru** [comment]
(Original issue was assigned to tshatch)

(Original issue 991 last updated on 2016-06-01T06:53:29.224157+00:00)

(Issue automaticaly closed due to status in Bitbucket: resolved)

**Comment by Anteru** [comment]
(Original comment by tshatch on 2014-05-20T14:21:50.846665+00:00)

Took a peek, yeah this is due to the dual-use of / for regex and division.

**Comment by Anteru** [comment]
(Original comment by lev_lozhkin on 2014-05-22T02:16:06.052850+00:00)

Here's a patch that allows proper highlighting of regex in this case.

```
#!diff
diff -r 2ba9b53c87ee pygments/lexers/web.py
--- a/pygments/lexers/web.py    Tue May 20 07:44:45 2014 -0700
+++ b/pygments/lexers/web.py    Wed May 21 19:05:18 2014 -0700
@@ -2469,21 +2469,23 @@
             (r'#{', String.Interpol, 'interpoling_string'),
             (r'[/#]', String.Regex),
         ],  
+        'operators': [
+            (r'\+\+|~|&&|\band\b|\bor\b|\bis\b|\bisnt\b|\bnot\b|\?|:|'
+             r'\|\||\\(?=\n)|'
+             r'(<<|>>>?|==?(?!>)|!=?|=(?!>)|-(?!>)|[<>+*`%&\|\^/])=?',
+             Operator),
+        ],
         'slashstartsregex': [
-            include('commentsandwhitespace'),
             (r'///', String.Regex, ('#pop', 'multilineregex')),
             (r'/(?! )(\\.|[^[/\\\n]|\[(\\.|[^\]\\\n])*])+/'
              r'([gim]+\b|\B)', String.Regex, '#pop'),
-            default('#pop'),
+            include('operators'), # consume the slash to avoid inf loop
+            default('#pop')
         ],
         'root': [
-            # this next expr leads to infinite loops root -> slashstartsregex
-            #(r'^(?=\s|/|<!--)', Text, 'slashstartsregex'),
             include('commentsandwhitespace'),
-            (r'\+\+|~|&&|\band\b|\bor\b|\bis\b|\bisnt\b|\bnot\b|\?|:|'
-             r'\|\||\\(?=\n)|'
-             r'(<<|>>>?|==?(?!>)|!=?|=(?!>)|-(?!>)|[<>+*`%&\|\^/])=?',
-             Operator, 'slashstartsregex'),
+            (r'(?=/)', Text, 'slashstartsregex'),
+            include('operators'),
             (r'(?:\([^()]*\))?\s*[=-]>', Name.Function),
             (r'[{(\[;,]', Punctuation, 'slashstartsregex'),
             (r'[})\].]', Punctuation),

```

Here are some test cases as well.

```
#!diff
diff -r 2ba9b53c87ee tests/examplefiles/example.coffee
--- a/tests/examplefiles/example.coffee Tue May 20 07:44:45 2014 -0700
+++ b/tests/examplefiles/example.coffee Wed May 21 18:55:59 2014 -0700
@@ -25,3 +25,34 @@
 '#wow'
 'wow#'
 'w#ow'
+
+# regex
+
+z = /a/.test(string)
+/\\/
+/\// # comment shouldnt be part of the regex
+/\\\//
+/(\s)/
+/a{2,8}/
+/b*c?d+/
+/(capture-match)/
+/(?:do-not-capture-match)/
+/this|or|that/
+/[char-set]/
+/[^neg-char_st]/
+/^.*$/
+/\n(\f)\0\1\d\b\cm\u1234/
+/^.?([^/\\\n\w]*)a\1+$/.something(or_other) # something more complex
+foo = (str) ->
+  /'|"/.test str
+a = a / b / c
+a = a/b/c
+a = a/b/ c
+a = a /b/c
+a = 1 + /d/.test(a)
+
+/// # multi-line regex
+(capture)
+[^excl-this] # a comment in multi-line regex
+///
```

All work as expected except a = a/b/ c. The /b/ is highlighted as regex there, any suggestions on how to fix that?

**Comment by Anteru** [comment]
(Original comment by tshatch on 2016-06-01T06:53:17.767770+00:00)

Rework CoffeeScript use of slashstartsregex.

Fixes #991

→ <<cset ae830aa036e8>>
